cpp/src/interop/arrow_data_structures.cpp

```diff
@@ -0,0 +1,347 @@
+ *
+ * The @p input must be the ``owner`` member of the @p container OR a child of
+ * the ``owner`` member of the @p container. If not, the behavior is undefined.
+ *
+ * @param output The ArrowArray to copy to
+ * @param input The ArrowArray to copy from
+ * @param container The container that owns the data
+ */
+void copy_array(ArrowArray* output,
+                ArrowArray const* input,
+                std::shared_ptr<arrow_array_container> container)
+{
+  auto private_data  = new ArrowArrayPrivateData{container};
+  output->length     = input->length;
+  output->null_count = input->null_count;
+  output->offset     = input->offset;
+  output->n_buffers  = input->n_buffers;
+  output->n_children = input->n_children;
+  output->buffers    = input->buffers;
+
+  if (input->n_children > 0) {
+    private_data->children_raw.resize(input->n_children);
+    for (auto i = 0; i < input->n_children; ++i) {
+      private_data->children.push_back(std::make_unique<ArrowArray>());
+      private_data->children_raw[i] = private_data->children.back().get();
+      copy_array(private_data->children_raw[i], input->children[i], container);
+    }
+  }
+  output->children     = private_data->children_raw.data();
+  output->dictionary   = input->dictionary;
+  output->release      = ArrayReleaseCallback;
+  output->private_data = private_data;
+}
+
+template <typename T>
+void arrow_obj_to_arrow(T& obj,
+                        std::shared_ptr<arrow_array_container> container,
+                        ArrowDeviceArray* output,
+                        ArrowDeviceType device_type,
+                        rmm::cuda_stream_view stream,
+                        rmm::device_async_resource_ref mr)
+{
+  switch (device_type) {
+    case ARROW_DEVICE_CUDA:
+    case ARROW_DEVICE_CUDA_HOST:
+    case ARROW_DEVICE_CUDA_MANAGED: {
+      auto& device_arr = container->owner;
+      copy_array(&output->array, &device_arr.array, container);
+      output->device_id = device_arr.device_id;
+      // We can reuse the sync event by reference from the input. The
+      // destruction of that event is managed by the destruction of
+      // the underlying ArrowDeviceArray of this table.
+      output->sync_event  = device_arr.sync_event;
+      output->device_type = device_type;
+      break;
+    }
+    case ARROW_DEVICE_CPU: {
+      auto out = cudf::to_arrow_host(*obj.view().get(), stream, mr);
+      ArrowArrayMove(&out->array, &output->array);
+      output->device_id   = -1;
+      output->sync_event  = nullptr;
+      output->device_type = ARROW_DEVICE_CPU;
+      break;
+    }
+    default: throw std::runtime_error("Unsupported ArrowDeviceArray type");
+  }
+}
+
+}  // namespace
+
+arrow_column::arrow_column(cudf::column&& input,
+                           column_metadata const& metadata,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr)
+  : container{[&]() {
```

Comment:
```suggestion
  : container{[&] {
```
The extra `()` is not necessary in IIFE.